radicle-node/src/service.rs

```diff
@@ -199,6 +200,10 @@ pub struct Service<R, A, S, G> {
     last_announce: LocalTime,
     /// Time when the service was initialized.
     start_time: LocalTime,
+
+    /// Storing the namespaces for fetches. Timestamp seems to be sufficient
+    /// to identify fetches at this point.
+    fetch_namespaces: HashMap<Timestamp, Namespaces>,
```

Comment:
Fetches are tracked in the peer session (`session.rs`) already. If you need to add namespace information you can add it there, but I'm not sure why you would need to..

Reply:
Did you mean the [`Protocol::Fetch`](https://github.com/radicle-dev/heartwood/blob/f055c9d4b38c0472dbc6843dbeeb9f7af8b9b5f0/radicle-node/src/service/session.rs#L30)?  This one is used to identify the state of the session protocol, not the fetch messages themselves.

When a peer receives `FetchOk` message, the session protocol state is actually [in `Gossip`](https://github.com/radicle-dev/heartwood/blob/f055c9d4b38c0472dbc6843dbeeb9f7af8b9b5f0/radicle-node/src/service.rs#L1147-L1151). It is clearer to carry the id (timestamp) in `FetchOk` message, instead of the protocol state.

Reply:
Right, see https://github.com/radicle-dev/heartwood/pull/370#discussion_r1133728229

All we need is to modify `Protocol::Gossip` slightly.